ts-client/src/amm/index.ts

```diff
@@ -0,0 +1,932 @@
+    const tokenMap = cluster === 'devnet' ? DEVNET_COIN : tokenListContainer.filterByClusterSlug(cluster).getList();
+
+    const tokenInfoA = tokenMap.find((token) => token.address === poolState.tokenAMint.toBase58());
+    const tokenInfoB = tokenMap.find((token) => token.address === poolState.tokenBMint.toBase58());
+    invariant(tokenInfoA, `TokenInfo ${poolState.tokenAMint.toBase58()} A not found`);
+    invariant(tokenInfoB, `TokenInfo ${poolState.tokenBMint.toBase58()} A not found`);
+
+    const vaultA = await VaultImpl.create(program.provider.connection, tokenInfoA, { cluster });
+    const vaultB = await VaultImpl.create(program.provider.connection, tokenInfoB, { cluster });
+
+    const poolInfo = await getPoolInfo({
+      poolMint: pool,
+      poolState,
+      program,
+      vaultA,
+      vaultB,
+      apyPda,
+    });
+
+    const depegAccounts = new Map<String, AccountInfo<Buffer>>();
+
+    if (poolState.curveType['stable'] && !poolState.curveType['stable']['depeg']['depegType']['none']) {
+      let extraAddresses: PublicKey[] = [];
+      for (const [, address] of Object.entries(CURVE_TYPE_ACCOUNTS)) {
+        extraAddresses = extraAddresses.concat(address);
+      }
+      const accounts = await program.provider.connection.getMultipleAccountsInfo(extraAddresses);
+      for (let i = 0; i < accounts.length; i++) {
+        const account = accounts[i];
+        const address = extraAddresses[i].toBase58();
+        if (account) {
+          depegAccounts.set(address, account);
+        }
+      }
+    }
+
+    const onChainTime = await getOnchainTime(program.provider.connection);
+
+    return new AmmImpl(
+      program,
+      apyPda,
+      [tokenInfoA, tokenInfoB],
+      pool,
+      { ...poolState, ...poolInfo },
+      vaultA,
+      vaultB,
+      depegAccounts,
+      onChainTime,
+      {
+        cluster,
+      },
+    );
+  }
+
+  get tokenA(): TokenInfo {
+    return this.tokenInfos[0];
+  }
+
+  get tokenB(): TokenInfo {
+    return this.tokenInfos[1];
+  }
+
+  get decimals(): number {
+    return Math.max(this.tokenA.decimals, this.tokenB.decimals);
+  }
+
+  get isStablePool(): boolean {
+    return 'stable' in this.poolState.curveType;
+  }
+
+  public async updatePoolState() {
+    const poolState = await getPoolState(this.address, this.program);
+    const poolInfo = await this.getPoolInfo();
+
+    this.poolState = { ...poolState, ...poolInfo };
+  }
+
+  /**
+   * It returns the pool token mint.
+   * @returns The poolState.lpMint
+   */
+  public getPoolTokenMint() {
+    return this.poolState.lpMint;
+  }
+
+  public async getLpSupply() {
+    const account = await this.program.provider.connection.getAccountInfo(this.poolState.lpMint);
+    invariant(account, ERROR.INVALID_ACCOUNT);
+
+    const lpMintInfo = MintLayout.decode(account.data);
+    const lpSupply = new BN(u64.fromBuffer(lpMintInfo.supply));
+    return lpSupply;
+  }
```

Comment:
this reason we have this async fn is because integrators can get the latest lpSupply. If they want it sync, they can access `AmmImpl.poolState.lpSupply`